File: api_nestjs_distributed_order_platform/src/application/use-case/product/get-product.use-case.ts
import { Inject, Injectable, NotFoundException } from "@nestjs/common";
import type { ProductRepositoryInterface } from "src/application/port/product.repository";
import type { CacheStoreInterface } from "src/application/port/redis-store.repository";
import { ProductEntity } from "src/domain/entities/product.entity";
import { PRODUCT_REPOSITORY } from "src/shared/tokens_nest/product.token";
import { CACHE_STORE } from "src/shared/tokens_nest/redis.token";

const PRODUCT_CACHE_TTL_SECONDS = 300;
@Injectable()
export class GetProductUseCase {
    constructor(
        @Inject(PRODUCT_REPOSITORY)
        private readonly productRepository: ProductRepositoryInterface,
        @Inject(CACHE_STORE)
        private readonly cacheStore: CacheStoreInterface,
    ) {}

    async execute(productId: string): Promise<ProductEntity> {
        const cacheKey = `product:${productId}`;

        const cached = await this.cacheStore.get<{
            sku: string;
            name: string;
            price: number;
            status: ProductEntity['getStatus'] extends () => infer S ? S : never;
            id: string;
        }>(cacheKey);

        if (cached) {
            return new ProductEntity(cached.sku, cached.name, cached.price, cached.status, cached.id);
        }

        const product = await this.productRepository.findById(productId);
        if (!product) {
            throw new NotFoundException('Product not found');
        }

        await this.cacheStore.set(cacheKey, {
            sku: product.getSku(),
            name: product.getName(),
            price: product.getPrice(),
            status: product.getStatus(),
            id: product.getId(),
        }, PRODUCT_CACHE_TTL_SECONDS);

        return product;
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/application/usecase/product/GetProductUseCase.java
package com.project.order.application.usecase.product;

import com.project.order.application.exception.NotFoundException;
import com.project.order.application.port.ProductRepository;
import com.project.order.application.port.RedisStoreRepository;
import com.project.order.domain.model.Money;
import com.project.order.domain.model.Product;
import com.project.order.domain.model.ProductStatus;
import org.springframework.stereotype.Service;

@Service
public class GetProductUseCase {

    private static final int PRODUCT_CACHE_TTL_SECONDS = 300;

    private record CachedProduct(String id, String sku, String name, double price, ProductStatus status) {}

    private final ProductRepository productRepository;
    private final RedisStoreRepository cacheStore;

    public GetProductUseCase(ProductRepository productRepository, RedisStoreRepository cacheStore) {
        this.productRepository = productRepository;
        this.cacheStore = cacheStore;
    }

    public Product execute(String productId) {
        String cacheKey = "product:" + productId;

        var cached = cacheStore.get(cacheKey, CachedProduct.class);
        if (cached.isPresent()) {
            CachedProduct c = cached.get();
            return new Product(c.id(), c.sku(), c.name(), new Money(c.price()), c.status());
        }

        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new NotFoundException("Product not found"));

        cacheStore.set(cacheKey, new CachedProduct(
                product.getId(), product.getSku(), product.getName(), product.getPrice().getValue(), product.getStatus()
        ), PRODUCT_CACHE_TTL_SECONDS);

        return product;
    }
}
